File: clients/python/tests/chunking/resources/kotlin/Sample.kt
/**
 * Sample Kotlin file for testing code chunking.
 *
 * This file contains various Kotlin constructs to test AST parsing.
 */
package sample

import kotlin.math.sqrt

// Constants
const val MAX_RETRIES = 3
const val DEFAULT_TIMEOUT = 30.0

/**
 * Represents a user in the system.
 */
data class User(
    val id: String,
    val name: String,
    var email: String? = null
) {
    fun getDisplayName(): String = name.ifEmpty { email ?: id }
}

/**
 * Custom exception for service errors.
 */
class ServiceException(message: String, val errorCode: String? = null) : Exception(message)

/**
 * Interface for services.
 */
interface Service {
    suspend fun initialize()
    fun isReady(): Boolean
}

/**
 * Sealed class for service results.
 */
sealed class Result<out T> {
    data class Success<T>(val data: T) : Result<T>()
    data class Error(val message: String, val code: String? = null) : Result<Nothing>()
}

/**
 * Enum for service status.
 */
enum class ServiceStatus {
    PENDING,
    RUNNING,
    STOPPED,
    ERROR
}

/**
 * Base class for services.
 */
abstract class BaseService(protected val config: Map<String, Any>) : Service {
    protected var initialized = false

    override suspend fun initialize() {
        initialized = true
    }

    override fun isReady(): Boolean = initialized

    protected fun validateConfig(): Boolean = config.isNotEmpty()
}

/**
 * Service for managing users.
 */
class UserService(config: Map<String, Any> = emptyMap()) : BaseService(config) {
    private val users = mutableMapOf<String, User>()

    fun createUser(id: String, name: String, email: String? = null): User {
        if (id.isEmpty()) {
            throw ServiceException("ID cannot be empty", "INVALID_ID")
        }
        val user = User(id, name, email)
        users[id] = user
        onUserCreated(user)
        return user
    }

    fun getUser(id: String): User? = users[id]

    fun deleteUser(id: String): Boolean = users.remove(id) != null

    fun getAllUsers(): List<User> = users.values.toList()

    private fun onUserCreated(user: User) {
        // Internal callback
    }
}

/**
 * Calculate factorial of n.
 */
fun calculateFactorial(n: Long): Long {
    return if (n <= 1) 1 else n * calculateFactorial(n - 1)
}

/**
 * Fetch data from URL (simulated).
 */
suspend fun fetchData(url: String, timeout: Double = DEFAULT_TIMEOUT): Map<String, String> {
    // Simulated fetch
    return mapOf("url" to url, "status" to "ok")
}

/**
 * Process items with optional validation.
 */
fun processItems(items: List<String>, validate: Boolean = true): List<String> {
    val filtered = if (validate) items.filter { it.isNotEmpty() } else items
    return filtered.map { it.trim().lowercase() }
}

/**
 * Extension function for String.
 */
fun String.truncate(maxLength: Int): String {
    return if (length <= maxLength) this else substring(0, maxLength)
}

/**
 * Object for utility functions.
 */
object MathUtils {
    fun fibonacci(n: Int): Long {
        return if (n <= 1) n.toLong() else fibonacci(n - 1) + fibonacci(n - 2)
    }

    fun distance(x: Double, y: Double): Double = sqrt(x * x + y * y)
}

/**
 * Inline function with reified type.
 */
inline fun <reified T> createInstance(): T? {
    return try {
        T::class.java.getDeclaredConstructor().newInstance()
    } catch (e: Exception) {
        null
    }
}

/**
 * Higher-order function.
 */
fun <T> withRetry(maxRetries: Int = MAX_RETRIES, block: () -> T): T {
    var lastException: Exception? = null
    repeat(maxRetries) {
        try {
            return block()
        } catch (e: Exception) {
            lastException = e
        }
    }
    throw lastException ?: ServiceException("Unknown error")
}

/**
 * Main entry point.
 */
suspend fun main() {
    val service = UserService(mapOf("env" to "test"))
    service.initialize()

    val user = service.createUser("1", "Test User", "test@example.com")
    println("Created user: ${user.getDisplayName()}")

    val result = calculateFactorial(5)
    println("Factorial: $result")
}

File: clients/python/tests/chunking/resources/swift/Sample.swift
/**
 * Sample Swift file for testing code chunking.
 *
 * This file contains various Swift constructs to test AST parsing.
 */

import Foundation

// MARK: - Constants

let maxRetries: Int = 3
let defaultTimeout: Double = 30.0

// MARK: - Custom Errors

enum ServiceError: Error {
    case invalidInput(String)
    case notFound(String)
    case internalError(String)
}

// MARK: - Protocols

protocol Service {
    func initialize() async throws
    var isReady: Bool { get }
}

protocol Displayable {
    var displayName: String { get }
}

// MARK: - User

/**
 * Represents a user in the system.
 */
struct User: Displayable, Codable, Equatable {
    let id: String
    let name: String
    var email: String?

    var displayName: String {
        if !name.isEmpty {
            return name
        }
        return email ?? id
    }

    init(id: String, name: String, email: String? = nil) {
        self.id = id
        self.name = name
        self.email = email
    }
}

// MARK: - Service Status

/**
 * Enum for service status.
 */
enum ServiceStatus {
    case pending
    case running
    case stopped
    case error(Error)
}

// MARK: - Base Service

/**
 * Base class for services.
 */
class BaseService: Service {
    let config: [String: Any]
    private(set) var initialized = false

    var isReady: Bool { initialized }

    init(config: [String: Any] = [:]) {
        self.config = config
    }

    func initialize() async throws {
        initialized = true
    }

    func validateConfig() -> Bool {
        return !config.isEmpty
    }
}

// MARK: - User Service

/**
 * Service for managing users.
 */
class UserService: BaseService {
    private var users: [String: User] = [:]

    func createUser(id: String, name: String, email: String? = nil) throws -> User {
        guard !id.isEmpty else {
            throw ServiceError.invalidInput("ID cannot be empty")
        }

        let user = User(id: id, name: name, email: email)
        users[id] = user
        onUserCreated(user)
        return user
    }

    func getUser(id: String) -> User? {
        return users[id]
    }

    func deleteUser(id: String) -> Bool {
        return users.removeValue(forKey: id) != nil
    }

    func getAllUsers() -> [User] {
        return Array(users.values)
    }

    private func onUserCreated(_ user: User) {
        // Internal callback
    }
}

// MARK: - Math Utilities

/**
 * Utility struct for math operations.
 */
struct MathUtils {
    /**
     * Calculate factorial of n.
     */
    static func calculateFactorial(_ n: UInt64) -> UInt64 {
        if n <= 1 { return 1 }
        return n * calculateFactorial(n - 1)
    }

    /**
     * Calculate fibonacci number.
     */
    static func fibonacci(_ n: Int) -> Int {
        guard n > 1 else { return n }
        return fibonacci(n - 1) + fibonacci(n - 2)
    }
}

// MARK: - Data Fetching

/**
 * Fetch data from URL asynchronously.
 */
func fetchData(url: String, timeout: Double = defaultTimeout) async throws -> [String: String] {
    // Simulated async fetch
    return ["url": url, "status": "ok"]
}

/**
 * Process items with optional validation.
 */
func processItems(_ items: [String], validate: Bool = true) -> [String] {
    var filtered = validate ? items.filter { !$0.isEmpty } : items
    return filtered.map { $0.trimmingCharacters(in: .whitespaces).lowercased() }
}

// MARK: - Generic Container

/**
 * Generic container class.
 */
class Container<T> {
    private var value: T

    init(_ value: T) {
        self.value = value
    }

    func get() -> T { value }
    func set(_ value: T) { self.value = value }
}

// MARK: - Extensions

extension String {
    func truncate(_ maxLength: Int) -> String {
        if count <= maxLength { return self }
        return String(prefix(maxLength))
    }
}

// MARK: - Typealiases

typealias UserId = String
typealias UserMap = [UserId: User]

// MARK: - Higher-order Functions

func withRetry<T>(maxRetries: Int = maxRetries, block: () throws -> T) rethrows -> T {
    var lastError: Error?
    for _ in 0..<maxRetries {
        do {
            return try block()
        } catch {
            lastError = error
        }
    }
    throw lastError!
}

// MARK: - Main

@main
struct Main {
    static func main() async {
        let service = UserService(config: ["env": "test"])
        do {
            try await service.initialize()

            let user = try service.createUser(id: "1", name: "Test User", email: "test@example.com")
            print("Created user: \(user.displayName)")

            let result = MathUtils.calculateFactorial(5)
            print("Factorial: \(result)")
        } catch {
            print("Error: \(error)")
        }
    }
}
